### PR TITLE
Simplify Condition Resolution

### PR DESCRIPTION
**Pull request by CodaFi**
This reverts the major content of #5778, the Condition Resolution pass, and replaces it with a far simpler implementation of condition resolution along with `canImport`.  When combined with the optimizations in #6279 we get the best of both worlds with a performance win and a simpler implementation.

This is one of two ways towards a stable implementation of this feature.  The concern here is that even with optimizations Parse will have to hit the disk at least once per `canImport` statement which could potentially slow down a very hot path.  On the other hand, that cost cannot simply be deferred to Sema without significantly complicating `DeclRef` resolution, semantic analysis, and condition resolution.

⚠️ Do Not Merge ⚠️ 

@benlangmuir, @jrose-apple thoughts?

Resolves [SR-3438](https://bugs.swift.org/browse/SR-3438), [SR-3544](https://bugs.swift.org/browse/SR-3544), and [SR-3546](https://bugs.swift.org/browse/SR-3546)

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
@swift-ci please test.

**Comment by CodaFi** [comment]
There.  @jrose-apple given the recent spate of issues, this is the version of this patch I'm most comfortable with.

File: validation-test/compiler_crashers_fixed/28611-isa-x-val-cast-ty-argument-of-incompatible-type.swift [hunk 1]
```diff
@@ -5,6 +5,6 @@
 // See https://swift.org/LICENSE.txt for license information
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
-// RUN: not --crash %target-swift-frontend %s -emit-ir
+// RUN: not %target-swift-frontend %s -emit-ir
 // REQUIRES: asserts
```
Comment (jrose-apple) [inline validation-test/compiler_crashers_fixed/28611-isa-x-val-cast-ty-argument-of-incompatible-type.swift]: No need for the REQUIRES: asserts.

**Comment by jrose-apple** [comment]
I guess this is the best we can do. Would still like @benlangmuir's opinion as well—it's annoying that we've added a language feature that means we *can't parse* without search paths set up. Maybe we can recover this later by taking @DougGregor's idea—the contents of active conditional blocks not having to be splatted into their parent. But the first priority is to restore source compat.

**Comment by benlangmuir** [comment]
I still strongly disagree with making parsing depend on (a) search paths, and (b) what is on disk in those search paths.  Not only is this a performance issue for syntax-only parsing, but also leaves us unable to reliably reproduce syntax-only SourceKit bugs without knowing the state of the filesystem or command-line argument, both of which are harder to find out and are less likely to be provided in bug reports than the source code.  We are stuck with that for semantic issues, but we shouldn't have to be for parsing.

**Comment by jrose-apple** [comment]
Then we're stuck with parse-only producing a slightly different AST for now. Or we turn the feature back off for 3.1 and give ourselves a little more time to plan.

**Comment by CodaFi** [comment]
Trouble is we're also stuck with it for Parse issues.  Parse depends on being able to have total knowledge of DeclRefs, classes introduced that depend on Foundation, and local type declarations in addition to structure in general.  Condition Resolution demonstrated that at least the last two could be shifted to Sema, but without ASTScope or NameLookup taking the first Parse is going to wire up nonsense.  Modeling conditions dynamically would make querying compound parts of the AST have to go through condition resolution which is still subject to at least one of the linked SRs (the one about deinitializers being especially concerning).

**Comment by jrose-apple** [comment]
The deinitializer one isn't so bad; the syntax-only parsing doesn't need to include members synthesized on demand. Sema can take care of that. We also don't actually care about resolving decl refs at all in a syntax-only parse.

But we're out of time. If @benlangmuir isn't happy with this approach, we should…well, we should *still* take this patch, so that we go back to what we had before, but we should disable the support for `canImport` until after Branch Day, and then think this through from the start again.

**Comment by benlangmuir** [comment]
> we should…well, we should still take this patch, so that we go back to what we had before, but we should disable the support for canImport until after Branch Day, and then think this through from the start again.

SGTM

**Comment by CodaFi** [comment]
OK, I'm going to remove support for the `canImport` feature, but add a requires annotation to the tests that are here for when we can work this out in the future.

**Comment by jrose-apple** [comment]
Thanks, Robert. I feel really bad that you've put all this work in (multiple implementations, effectively!) and we can't ship it yet. Ben and I talked a little more about what we think the syntax-only parser wants out of this and I'll send you and swift-dev an email on it later.

**Comment by CodaFi** [comment]
Don't worry about it, I'm most concerned about keeping the feature and language stable around the change.

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
Due to #6631 

@swift-ci please smoke test and merge.

**Comment by CodaFi** [comment]
I can't find the test anywhere, so 

@swift-ci please smoke test Linux platform.

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**